README.md
# VoskJs

VoskJs is a NodeJs developers toolkit to use [Vosk](https://alphacephei.com/vosk/) 
offline speech recognition engine, including multi thread (server) usage examples. 
The project gives you: 

- A simple sentence-based and streaming-based transcript APIs
- The command line utility `voskjs`
- A demo HTTP transcript server `voskjshttp`

VoskJs can be used for speech recognition processing in different scenarios:

- Single-user/standalone programs (e.g. perfect for single-user embedded systems) 
- Multi-user/multi-core server architectures 


## What's Vosk?

Vosk is an open source embedded (offline/on-prem) speech-to-text engine 
which can run with very low latencies (`< 500`msecs on my [PC](tests/README.md#my-hardware--host-configuration)).
Vosk is based on a common DNN-HMM architecture.  Deep neural network is used for sound scoring (acoustic scoring), 
HMM and WFST frameworks are used for time models (language models).
It's based on [Kaldi](https://github.com/kaldi-asr/kaldi),
but Nikolay V. Shmyrev's Vosk offers a smart, simplified and performing interface! 
More details in the [Vosk home page](https://alphacephei.com/vosk/) 
and [github repo](https://github.com/alphacep/vosk-api).


## What's VoskJs?

The goal of the project is to create an simple function API layer 
on top of already existing Vosk nodejs binding, 
supplying both sentence-based and streaming-based speech-to-text functionalities.

### Sentence-based transcript API

In this mode, a file or a PCM buffer are processed asynchronously, 
to get the full text transcript of the given speech. 
Using the simple transcript interface you can build your standalone custom application, 
accessing async functions suitable to run on a usual single thread nodejs program.

Pseudo code:

```javascript
//Loads once in RAM memory a specific Vosk engine model from a model directory.
const model = loadModel(modelDirectory)

// transcripts a speech file or buffer (in WAV/PCM format), using Vosk engine. 
// It supply speech-to-text transcript detailed info.
const result = await transcriptFromFile(fileName, model, {options}) 

// or 
// const result = await transcriptFromBuffer(buffer, model, {options}) 

freeModel(model)
```

### Streaming-based transcript API (DRAFT)

Following Vosk-api recognizer result functions, VoskJs emit these nodejs events:

| Event name    | Vosk-api recognizer function | description                                 |
| ------------- | ---------------------------- | ------------------------------------------- |
| `partial`     | recognizer.patialResult()    | silent (text = '') or new word or new words |
| `endOfSpeech` | recognizer.result()          | end of speech (words followed by a silence) |
| `final`       | recognizer.finalResult()     | last part of the audio                      |

Pseudo code:

```javascript
//Loads once in RAM memory a specific Vosk engine model from a model directory.
const model = loadModel(modelDirectory)

const transcriptEvents = transcriptEventsFromFile(fileName, model, {options}) 
// or
// const transcriptEvents = transcriptEventsFromBuffer(buffer, model, {options}) 

// an new word is detected
transcriptEvents.on('partial', data => console.log(data) ) 

// a complete sentence (followed by silence) is detected 
transcriptEvents.on('endOfSpeech', data => console.log(data) )

// final (last) sentence is detected
transcriptEvents.on('final', data => console.log(data) )

freeModel(model)
```


### Command line tools

- `voskjs`: command line program to test Vosk transcript with specific models 
  (some tests and command line usage [here](tests/README.md)).

  BTW the utility can be configured to tabularize events. By example:

  ```
  voskjs --audio=audio/sentencesWithSilences.wav --model=models/vosk-model-small-en-us-0.15 --tableevents
  ```
  ```
  voskjs is a CLI utility to test Vosk-api features
  package @solyarisoftware/voskjs version 1.2.7, Vosk-api version 0.3.30

  Statistics:

  model directory      : models/vosk-model-small-en-us-0.15
  speech file name     : audio/sentencesWithSilences.wav
  grammar              : not specified. Default: NO
  sample rate          : not specified. Default: 16000
  max alternatives     : undefined
  text only / JSON     : JSON
  Vosk debug level     : -1

  load model latency   : 2001ms
  transcript latency   : 1707ms
  transcript text      : one two three four five six seven eight nine zero one two three stop 

  Events table:

  | time   | event        | text                                     |
  | ------ | ------------ | ---------------------------------------- |
  |     66 | partial      | 
  |    489 | partial      | one
  |    538 | partial      | one two
  |    592 | partial      | one two three
  |    635 | endOfSpeech  | one two three
  |    668 | partial      | 
  |    847 | partial      | for
  |    882 | partial      | four five six
  |    977 | partial      | four five six seven
  |   1099 | partial      | four five six seven eight
  |   1169 | endOfSpeech  | four five six seven eight
  |   1194 | partial      | 
  |   1322 | partial      | nine
  |   1381 | partial      | nine zero
  |   1456 | partial      | nine zero one
  |   1498 | partial      | nine zero one two
  |   1550 | partial      | nine zero one two three
  |   1630 | partial      | nine zero one two three stop
  |   1649 | endOfSpeech  | nine zero one two three stop
  |   1677 | partial      | 
  |   1706 | final        | 
  ```

- `voskjshttp`: a simple demo HTTP server to transcript speech files. 
  Using above API you can build your own server. Some usage examples [here](examples/).


## 🛍 Install 

### 1. Install Vosk engine and this nodejs module 

- Install vosk-api engine
  ```bash
  pip3 install -U vosk 
  ```
  See also: https://alphacephei.com/vosk/install

- Install this module, as global package if you want to use CLI command `voskjs` 
  ```bash
  npm install -g @solyarisoftware/voskjs@latest
  ```


### 2. Install/Download Vosk models

```bash
mkdir your/path/models && cd models

# English large model
wget https://alphacephei.com/vosk/models/vosk-model-en-us-aspire-0.2.zip
unzip vosk-model-en-us-aspire-0.2.zip

# English small model
wget http://alphacephei.com/vosk/models/vosk-model-small-en-us-0.15.zip
unzip vosk-model-small-en-us-0.15.zip

# Italian model model
wget https://alphacephei.com/vosk/models/vosk-model-small-it-0.4.zip
unzip vosk-model-small-it-0.4.zip
```

More about available Vosk models here: https://alphacephei.com/vosk/models

### 3. Demo audio files

Directory [`audio`](audio/) contains some English language speech audio files, 
coming from a Mozilla DeepSpeech repo.
Source: [Mozilla DeepSpeech audio samples](https://github.com/mozilla/DeepSpeech/releases/download/v0.9.3/audio-0.9.3.tar.gz)
These files are used for some tests and comparisons.


## 🧐 [Examples](examples/README.md)

Some VoskJs usage examples:

- [Simple program for a sentence-based speech-to-text](examples/README.md#simple-program-for-a-sentence-based-speech-to-text)
- [`voskjs` Command line utility](examples/README.md#voskjs-command-line-utility)
- [`voskjshttp` demo speech-to-text HTTP server](examples/servers.md#voskjshttpjs-demo-spech-to-text-http-server)
- [`voskjshttp` as RHASSPY speech-to-text remote HTTP Server](examples/servers.md#voskjshttp-as-rhasspy-speech-to-text-remote-http-server)
- [Sentence-based speech-to-text, specifying a grammar](examples/grammars.md#sentence-based-speech-to-text-specifyng-a-grammar)
- [SocketIO server pseudocode](examples/servers.md#socketio-server-pseudocode)


## 🛠 [Tests](tests/README.md)

Some tests/notes:

- Transcript using English language, large model
- Transcript using English language, small model
- Comparison between Vosk and Mozilla DeepSpeech (latencies)
- Multi-thread stress test (10 requests in parallel)
- HTTP Server benchmark test
- Latency tests


## 🎁 Bonus track

[`audioutils`](lib/audioutils.js) some audio utility functions as `toPCM`, 
a fast transcoding to PCM, using ffmpeg process (install ffmpeg before).


## To do

- To speedup latencies, rethink transcript interface, maybe with an initialization phases, 
  including Model and Recognizer(s) object creation.
  Possible architecture: [Stateful & low latency ASR architecture](https://github.com/alphacep/vosk-api/issues/553)
- Deepen grammar usage with more examples
- Deepen Vosk-API errors catching
- `voskjshttp`: 
  - Review stress and performances tests (especially for the HTTP server)
  - HTTP POST management: 
    - set mandatory audio format mime type in the header request (`--header "Content-Type: audio/wav"`)
    - audio-transcoding using function `toPcm` 
      if input speech files are not specified as wav in header request (e.g. `--header "Content-Type: audio/webm"`)
      see https://cloud.ibm.com/docs/speech-to-text?topic=speech-to-text-audio-formats#audio-formats-list


## How to contribute


CHANGELOG.md
# Change log

- 1.2.8 
  - duplicated `partial` events are no more emitted
  - README.md updated

- 1.2.6 
  - voskjs new flags for pretty printing of streaming events: `tableevents`, `objectevents`

- 1.2.4
  - voskjs.js modified to better manage streaming events `partial`, `endOfSpeech` and `final`
  - voskjs.js: added `transcriptEventsFromFile` to manage streaming transcript with events. Just a draft
  - Now voskjs prints partial results, end of speech results and final result.
  - new audio file for test: audio/sentencesWithSilences.wav

- 1.1.5
  - docs update
  - updated to integrate vosk package 0.3.30 new functions (setAlternatives, setWords)

- 1.1.1
  - voskjshttp JSON response attribute *result* renamed *vosk*
  - lib/audioutils.js substitute lib/toPCM.js
  - httpServer.js renamed voskjshttp.js
  - documenattion updated

- 1.0.8
  - added flag `--no-threads` to `voskjshttp`
  - added voskjshttp docs to explain how to use it as a RHASSPY speech-to-text HTTP Remote Server 
  - documentation update

- 1.0.3
  - voskjs take sample rate as command line parameter
  - voskjshttp updated 
    - reply application/json or text/plain following request header attribute "Accept"
    - endpoint path /transcript can be configured 
    - added test scripts clientGETtext.sh and clientPOSTtext.sh
  - documentation update

- 1.0.2
  - added versions info in voskjs and voskjshttp
  - clientGET*.sh scripts updated using a full path name for the WAV file
  - httpserver check contet-type 

- 1.0.0
  - functions loadModule, transcriptFromFile, transcriptFromBuffer now returns simple values instead of objects containing latency attribute
  - added module lib/chronos.js to manage latencies 
  - added createRecognizer function prepared to manager MaxAlternatives
  - added flag --debug to voskjs cli

- 0.5.2
  - lib/toPCM updated
  - Documentaion rework

- 0.5.1
  - httpServer.js is now renamed voskjshttp.js. It accepts GET and POST endpoints
  - voskjshttp now reply to an HTTP POST request receiving the speech WAV 
    file as binary data attached in the HTTP request:
    ```
     curl -X POST 'http:localhost:3000/transcript' \
      --header "Content-Type: audio/wav" \
      --data-binary "@speech.wav"
    ```

- 0.4.0
  - httpServer interface changed. Now accepting HTTP GET requests and query string arguments.

- 0.3.14
  - added latency tests (with/without grammars)
  - added function `transcriptFromBuffer` that transcripts audio as buffer instead of as a WAV file
  - added `lib/toPCM.js` module function to manage ffmpeg input/output Buffers, 
    through stdin/stdout, avoiding disk I/O

- 0.3.11
  - added some tests to verify ffmpeg transcoding times 
    (using stdout pipe / without producing wav file)

- 0.3.8
  - Integrate CLI command `voskjs` with grammar argument 
  - global installation of the package allow to run `voskjs` cli program
  - socketio server program example

- 0.3.6
  - delivered @solyarisoftware/voskjs (added index.js)
  - httpServer now accepts grammar attribute in client API calls 
  - added test program tests/sequentialRequests.js
  - function `transcript` can be called with alias `recognize`
  - function `transcript` now accept options argument, allowing to manage Vosk grammar format.
  - documenation improved (README, examples/README, tests/README) 

- 0.2.11
  - added com/abtest.sh, apache bench httpServer stress test script
  - httpServer improved: 
    - better validation of body request 
    - accept optionally client request id as attribute in the request body
    - takes model, port and debug attributes as command line args.
  - documenation improved (README, examples/README, tests/README) 

- 0.2.1
  - added function logLevel
  - httpServer error responses improvements

- 0.2.0
  - function initModel is renamed loadModel
  - function loadModel and Transcript  return an object containing also the processing latency
  - httpServer reviewed. The transcript endpoint return a different JSON data structure, containing latency time.

- 0.1.0 
  Added a simple HTTP sever

- 0.0.15 
  Added tests directory, containing some stress tests results

- 0.0.14 
  Transcript function updated to integrate Vosk version 0.3.25 (`npm install vosk@latest`), 
  where the function `rec.acceptWaveformAsync` now run on a separated external thread!

---

[top](#) | [home](README.md)



voskjs.js
#!/usr/bin/env node

/**
 * @module voskjs 
 *
 * @public
 * @function logLevel
 * @function loadModel
 * @function transcript
 * @function freeModel
 *
-* @see VoskAPI https://github.com/alphacep/vosk-api/blob/master/nodejs/index.js
 * @see https://github.com/alphacep/vosk-api/blob/master/nodejs/index.js#L207
 * @see https://www.tutorialsteacher.com/nodejs/nodejs-eventemitter
 * util.inherits(vosk.Recognizer, emitter)
 */ 

const fs = require('fs')
const util = require('util')
const emitter = require('events').EventEmitter

const { Readable } = require('stream')
const wav = require('wav')
const vosk = require('vosk')

const { info } = require('./lib/info')
const { getArgs } = require('./lib/getArgs')
const { setTimer, getTimer, unixTimeMsecs } = require('./lib/chronos')


/**
 * @constant
 */
const SAMPLE_RATE = 16000

const PARTIAL_RESULT_EVENT = 'partial' 
const END_OF_SPEECH_EVENT = 'endOfSpeech' 
const FINAL_RESULT_EVENT = 'final' 


/**
 * @function logLevel
 * @public
 * Set log level for Vosk/Kaldi log messages
 *
 * @param {number} level   The higher, the more verbose. 0 for infos and errors. Less than 0 for silence. 
 */
function logLevel(level=0) {

  // set vosk log level
  vosk.setLogLevel(level)

}  


/**
 * @function loadModel 
 * Create a run time model from the specified directory 
 *
 * @public
 *
 * @param {String}       modelDirectory directory name of the Vosk model
 *
 * @typedef ModelObject
 * @property {VoskModel} model          run time model object returned by Vosk engine.
 * @property {Number}    latency        elpased time in msecs
 *
 * @return {promise<ModelObject>}
 *
 */ 
function loadModel(modelDirectory) {

  // check if model directory exists, async
  fs.access(modelDirectory, (err) => {
    if (err) 
      throw `${err}: file ${modelDirectory} not found.`
  })
  
  // create new run time model from the specified directory 
  // TODO try/catch?
  const model = new vosk.Model(modelDirectory)

  return model 

}  


/**
 * @function createRecognizer 
 *
 * Create a run time Vosk recognizer 
 *
 * @typedef VoskRecognizerArgsObject
 * @property {Boolean}                 multiThreads if true, an external (Vosk engine) thread is spawned on the fly
 *                                                  that need in server (concurrent requests) architecture.
 * @property {Number}                  sampleRate   Default value: 16000
 * @property {String[]}                grammar      array of words, or sentences
 * @property {Number}                  alternatives maximum alternatives to return from recognition results
 * @property {Boolean}                 words        if true, recognizer result will include word by word details
 * @example
 *   {
 *     result: [
 *       { conf: 1, end: 1.02, start: 0.36, word: 'experience' },
 *         { conf: 1, end: 1.35, start: 1.02, word: 'proves' },
 *        { conf: 1, end: 1.74, start: 1.35, word: 'this' }
 *      ],
 *      text: 'experience proves this'
 *    }
 *
 * @param {ModelObject}                model        the Vosk model returned by InitModel()
 * @param {VoskRecognizerArgsObject}  [options]     Vosk Recognizer arguments setting. Optional
 * @return {VoskRecognizerObject}
 *
 */ 
function createRecognizer(model, { sampleRate=SAMPLE_RATE, grammar=null, alternatives=0, words=true } = {}) {

  // if a grammar is specified, pass it to the Vosk Recognizer
  const voskRecognizerArgs = grammar ? 
    {model, sampleRate, grammar} :
    {model, sampleRate}

  // create Vosk recognizer
  // TODO try/catch?
  const recognizer = new vosk.Recognizer(voskRecognizerArgs)

  if ( alternatives )
    recognizer.setMaxAlternatives(alternatives)

  recognizer.setWords(words)

  return recognizer
}


/**
 * @function transcriptFromFile
 * speech recognition into a text, from an audio file, given a specified Vosk model
 *
 * @alias transcript
 * @public
 * @async
 *
 * @param {String}                     fileName     the name of speech file, in WAV format
 * @param {ModelObject}                model        the Vosk model returned by InitModel()
 * @param {VoskRecognizerArgsObject}   [options]    Vosk Recognizer arguments setting. Optional. 
 *
 * @return {Promise<VoskResultObject>} transcript object returned by Vosk engine
 *
 */ 
async function transcriptFromFile(fileName, model, { multiThreads=true, sampleRate=SAMPLE_RATE, grammar=null, alternatives=0, words=true } = {}) {

  const DEBUG = false

  return new Promise( (resolve, reject) => {

    // validate audiofile existence, async
    fs.access(fileName, (err) => {
      if (err) 
        return reject(`${err}: file ${fileName} not found.`)
    })

    if (DEBUG)
      setTimer('createRecognizer')

    const recognizer = createRecognizer( model, {sampleRate, grammar, alternatives, words} )

    if (DEBUG)
      console.log(`recognizer latency   : ${getTimer('createRecognizer')}ms`)

    const wfStream = fs.createReadStream(fileName, {'highWaterMark': 4096})
    const wfReader = new wav.Reader()
    
    wfStream.pipe(wfReader)
    
    const pcmChunks = new Readable().wrap(wfReader)

    wfReader.on('format', async ( { audioFormat, sampleRate, channels } ) => {
        
      if (audioFormat != 1 || channels != 1)
        return reject(`${fileName}: audio file (sample rate: ${sampleRate}) must be WAV format mono PCM.`)


      for await (const data of pcmChunks) {

        //
        // WARNING
        // From vosk version 0.3.25
        // the acceptWaveformAsync function runs in a dedicated thread.
        // That wold improve performances in case of concurrent requests 
        // from the caller (server) program  
        //
        // Previous vosk version 0.3.25
        // const end_of_speech = recognizer.acceptWaveform(data)
        //
        const end_of_speech = multiThreads ? 
          await recognizer.acceptWaveformAsync(data) : 
          recognizer.acceptWaveform(data)

        //
        // WARNING
        // 1. AcceptWaveform returns true when silence is detected and you can retrieve the result with Result(). 
        // 2. If silence is not detected you can retrieve PartialResult() only. 
        // 3. FinalResult means the stream is ended, you flush the buffers and retrieve remaining result.
        // By Nicolay Shmirev. See: https://github.com/alphacep/vosk-api/issues/590#issuecomment-863065813
        //
        if (end_of_speech) {
          // End of speech means silence detected.
          // We want to transcript all the audio so the processing continue until the end.

          // debug
          //console.log('DEBUG', END_OF_SPEECH_EVENT, recognizer.result())
          continue

        }
        //else
        // console.log('partialResult', recognizer.partialResult())
        
      }

      // copy final Vosk engine result object
      const result = {...recognizer.finalResult(recognizer)} 

      recognizer.free()
      
      return resolve(result)

    })
  })

}


/**
 * @function transcriptEventsFromFile
 *
 * speech recognition into a text, from an audio file, given a specified Vosk model,
 * return an events emitter
 *
 * @public
 *
 * @param {String}                     fileName     the name of speech file, in WAV format
 * @param {ModelObject}                model        the Vosk model returned by InitModel()
 * @param {VoskRecognizerArgsObject}   [options]    Vosk Recognizer arguments setting. Optional. 
 *
 * @return {Emitter}                   emit events
 *
 */ 
function transcriptEventsFromFile(fileName, model, { multiThreads=true, sampleRate=SAMPLE_RATE, grammar=null, alternatives=0, words=true } = {}) {

  const DEBUG = false

  // validate audiofile existence, async
  fs.access(fileName, (err) => {
    if (err) 
      throw (`${err}: file ${fileName} not found.`)
  })

  if (DEBUG)
    setTimer('createRecognizer')

  // the function is enabled to emit events 
  const event = new emitter()

  const recognizer = createRecognizer( model, {sampleRate, grammar, alternatives, words} )

  if (DEBUG)
    console.log(`recognizer latency   : ${getTimer('createRecognizer')}ms`)

  const wfStream = fs.createReadStream(fileName, {'highWaterMark': 4096})
  const wfReader = new wav.Reader()
  
  wfStream.pipe(wfReader)
  
  const pcmChunks = new Readable().wrap(wfReader)

  wfReader.on('format', async ( { audioFormat, sampleRate, channels } ) => {
      
    if (audioFormat != 1 || channels != 1)
      throw (`${fileName}: audio file (sample rate: ${sampleRate}) must be WAV format mono PCM.`)

    const lastPartialResult = {}

    for await (const data of pcmChunks) {

      //
      // WARNING
      // From vosk version 0.3.25
      // the acceptWaveformAsync function runs in a dedicated thread.
      // That wold improve performances in case of concurrent requests 
      // from the caller (server) program  
      //
      // Previous vosk version 0.3.25
      // const end_of_speech = recognizer.acceptWaveform(data)
      //
      const end_of_speech = multiThreads ? 
        await recognizer.acceptWaveformAsync(data) : 
        recognizer.acceptWaveform(data)

      //
      // WARNING
      // Emit partial result events
      // 1. AcceptWaveform returns true when silence is detected and you can retrieve the result with Result(). 
      // 2. If silence is not detected you can retrieve PartialResult() only. 
      // 3. FinalResult means the stream is ended, you flush the buffers and retrieve remaining result.
      // By Nicolay Shmirev. See: https://github.com/alphacep/vosk-api/issues/590#issuecomment-863065813
      //
      if (end_of_speech)
        event.emit(END_OF_SPEECH_EVENT, recognizer.result())
      else {
        const partialResult = recognizer.partialResult()

        // Doesn't emit duplicated events:
        // Is the current partialResult different form the last one?
        // If true a new event is emiitted.
        // This avoid "duplicated" events to be emitted
        if (partialResult.partial !== lastPartialResult.partial) {
          event.emit(PARTIAL_RESULT_EVENT, partialResult)
          lastPartialResult.partial = partialResult.partial
        }
      }
    }  
    
    //
    // WARNING
    // Emit partial result events
    // 1. AcceptWaveform returns true when silence is detected and you can retrieve the result with Result(). 
    // 2. If silence is not detected you can retrieve PartialResult() only. 
    // 3. FinalResult means the stream is ended, you flush the buffers and retrieve remaining result.
    // By Nicolay Shmirev. See: https://github.com/alphacep/vosk-api/issues/590#issuecomment-863065813
    //
    event.emit(FINAL_RESULT_EVENT, recognizer.finalResult(recognizer)) 

    recognizer.free()
    
  })

  return event

}


/**
 * @function transcriptFromBuffer
 * speech recognition into a text, from an audio file, given a specified Vosk model
 *
 * @alias transcript
 * @public
 * @async
 *
 * @param {Buffer}                     buffer       input buffer, in PCM format
 * @param {ModelObject}                model        the Vosk model returned by InitModel()
 * @param {VoskRecognizerArgsObject}   [options]    Vosk Recognizer arguments setting. Optional. 
 *
 * @return {Promise<VoskResultObject>} transcript object returned by Vosk engine
 *
 */ 
async function transcriptFromBuffer(buffer, model, { multiThreads=true, sampleRate=SAMPLE_RATE, grammar=null, alternatives=0, words=true } = {}) {

  const recognizer = createRecognizer( model, {sampleRate, grammar, alternatives, words} )
   
  // https://gist.github.com/wpscholar/270005d42b860b1c33cf5ab25b37928a
  // https://stackoverflow.com/questions/47089230/how-to-convert-buffer-to-stream-in-nodejs
  
  //
  // WARNING
  // From vosk version 0.3.25
  // the acceptWaveformAsync function runs in a dedicated thread.
  // That wold improve performances in case of cocurrent requests 
  // from the caller (server) program  
  //
  // Previous vosk version 0.3.25
  // const end_of_speech = recognizer.acceptWaveform(data)
  //
  if ( multiThreads ) 
    await recognizer.acceptWaveformAsync(buffer)
  else
    recognizer.acceptWaveform(buffer)

  // copy final Vosk engine result object
  const result = {...recognizer.finalResult(recognizer)} 

  recognizer.free()
    
  return Promise.resolve(result)

}


/**
 * @function transcriptEventsFromBuffer
 * speech recognition into a text, from an audio file, given a specified Vosk model
 *
 * @alias transcript
 * @public
 * @async
 *
 * @param {Buffer}                     buffer       input buffer, in PCM format
 * @param {ModelObject}                model        the Vosk model returned by InitModel()
 * @param {VoskRecognizerArgsObject}   [options]    Vosk Recognizer arguments setting. Optional. 
 *
 * @return {Emitter}                                
 *
 */ 
async function transcriptEventsFromBuffer(buffer, model, { multiThreads=true, sampleRate=SAMPLE_RATE, grammar=null, alternatives=0, words=true } = {}) {

  // the function is enabled to emit events 
  const event = new emitter()

  const recognizer = createRecognizer( model, {sampleRate, grammar, alternatives, words} )
   
  // https://gist.github.com/wpscholar/270005d42b860b1c33cf5ab25b37928a
  // https://stackoverflow.com/questions/47089230/how-to-convert-buffer-to-stream-in-nodejs
  
  const lastPartialResult = {}

  for(;;) {

    //
    // WARNING
    // From vosk version 0.3.25
    // the acceptWaveformAsync function runs in a dedicated thread.
    // That wold improve performances in case of cocurrent requests 
    // from the caller (server) program  
    //
    // Previous vosk version 0.3.25
    // const end_of_speech = recognizer.acceptWaveform(data)
    //
    const end_of_speech = multiThreads ? 
      await recognizer.acceptWaveformAsync(buffer) : 
      recognizer.acceptWaveform(buffer)

    //
    // WARNING
    // Emit partial result events
    // 1. AcceptWaveform returns true when silence is detected and you can retrieve the result with Result(). 
    // 2. If silence is not detected you can retrieve PartialResult() only. 
    // 3. FinalResult means the stream is ended, you flush the buffers and retrieve remaining result.
    // By Nicolay Shmirev. See: https://github.com/alphacep/vosk-api/issues/590#issuecomment-863065813
    //
    if (end_of_speech)  {
      event.emit(END_OF_SPEECH_EVENT, recognizer.result())
      continue
    }
    else {
      const partialResult = recognizer.partialResult()

      // Doesn't emit duplicated events:
      // Is the current partialResult different form the last one?
      // If true a new event is emiitted.
      // This avoid "duplicated" events to be emitted
      if (partialResult.partial !== lastPartialResult.partial) {
        event.emit(PARTIAL_RESULT_EVENT, partialResult)
        lastPartialResult.partial = partialResult.partial
      }

      break
    }  
  }

  //
  // WARNING
  // if end_Of_Speech is detected (the buffere contains a sentence followed by a silence)
  // and the result() function is called, so
  // the finalResult() contains just the remaining (last) part of the sentence before the end of the audio.
  // It's up to user to collect events data assempling the final textual (multisentence) result.
  //
  event.emit(FINAL_RESULT_EVENT, recognizer.finalResult())

  recognizer.free()
    
  return event

}



/**
 * @function freeModel
 * @public
 *
 * @param {ModelObject} model
 *
 */ 
function freeModel(model) {
  model.free()
}  


/**
 * test section
 */

function helpAndExit() {
  console.log('voskjs is a CLI utility to test Vosk-api features')
  console.log (info())
  console.log()
  console.log('Usage')
  console.log()
  console.log('  voskjs \\ ')
  console.log('    --model=<model directory> \\ ')
  console.log('    --audio=<audio file name> \\ ')
  console.log('    [--grammar=<list of comma-separated words or sentences>] \\ ')
  console.log('    [--samplerate=<Number, usually 16000 or 8000>] \\ ')
  console.log('    [--alternatives=<number of max alternatives in text result>] \\ ')
  console.log('    [--textonly] \\ ')
  console.log('    [--tableevents] \\ ')
  console.log('    [--objectevents] \\ ')
  console.log('    [--debug=<Vosk debug level>] ')
  console.log()    
  console.log('Examples')
  console.log()
  console.log('  1. Recognize a speech file using a specific model directory:')
  console.log()
  console.log('     voskjs --audio=audio/2830-3980-0043.wav --model=models/vosk-model-en-us-aspire-0.2')
  console.log()
  console.log('  2. Recognize a speech file setting a grammar (with a dynamic graph model) and a number of alternative:')
  console.log()
  console.log('     voskjs \\ ')
  console.log('       --audio=audio/2830-3980-0043.wav \\ ')
  console.log('       --model=models/vosk-model-small-en-us-0.15 \\ ')
  console.log('       --grammar="experience proves this, bla bla bla"')
  console.log('       --alternatives=3')
  console.log()
  process.exit(1)
}  


/**
 * @function checkArgs
 * command line parsing
 *
 * @param {String}    args
 *
 * @typedef {Object}  SentenceAndAttributes
 * @property {String} language 
 * @returns {SentenceAndAttributes}
 * 
 */
function checkArgs(args) {

  // mandatory arguments
  const modelDirectory = args.model 
  const audioFile = args.audio 

  // optional arguments
  const grammar = args.grammar 
  const sampleRate = args.samplerate 
  const alternatives = args.alternatives 
  const textOnly = args.textonly 
  const tableevents = args.tableevents 
  const objectevents = args.objectevents 


  // if not specified, set default Vosk debug level to -1 (silent mode)
  const debug = args.debug ? args.debug : -1

  if ( !modelDirectory ) 
    helpAndExit()

  if ( !audioFile ) 
    helpAndExit()

  return { 
    modelDirectory, 
    audioFile, 

    // if grammar args is present, as comma separated sentences,
    // convert it in an array of strings
    grammar: grammar ? grammar.split(',').map(sentence => sentence.trim()) : undefined, 

    // convert to Number
    sampleRate: sampleRate ? +sampleRate : undefined,

    alternatives,
    textOnly, 
    tableevents, 
    objectevents, 
    debug
  }
}


function printObject(object, args) {
   const defaultArgs = {showHidden:false, breakLength:Infinity, depth:null, colors:true}
   return util.inspect(object,  {...defaultArgs, ...args } )
}  


function printResultsAsTable(results) {

  console.log ('Events table:')
  console.log ()

  console.log('| %s | %s | %s |',
    'time'.padEnd(6), 
    'event'.padEnd(12), 
    'text'.padEnd(40)
  )
  console.log('| %s | %s | %s |',
    '-'.repeat(6), 
    '-'.repeat(12), 
    '-'.repeat(40)
  )

  for (const result of results) {
  
    if ( result.event === PARTIAL_RESULT_EVENT )
      console.log('| %s | %s | %s',
        result.time.toString().padStart(6),
        result.event.padEnd(12),
        result.data.partial //.padStart(11+6+2+1)
      )

    if ( result.event === END_OF_SPEECH_EVENT )
      console.log('| %s | %s | %s',
        result.time.toString().padStart(6),
        result.event.padEnd(12),
        result.data.text //.padStart(11+6+2+1)
      )
    
    if ( result.event === FINAL_RESULT_EVENT )
      console.log('| %s | %s | %s',
        result.time.toString().padStart(6),
        result.event.padEnd(12),
        result.data.text //.padStart(11+6+2+1)
      )
  }

  console.log()

}  


function statistics(modelDirectory, audioFile, grammar, sampleRate, alternatives, textOnly, tableevents, objectevents, debug, sentences) {

  console.log('voskjs is a CLI utility to test Vosk-api features')
  console.log (info())
  console.log()
  console.log('Statistics:')
  console.log()
  console.log(`model directory      : ${modelDirectory}`)
  console.log(`speech file name     : ${audioFile}`)
  console.log(`grammar              : ${grammar ? grammar : 'not specified. Default: NO'}`)
  console.log(`sample rate          : ${sampleRate ? sampleRate : 'not specified. Default: 16000'}`)
  console.log(`max alternatives     : ${alternatives}`)
  console.log(`text only / JSON     : ${textOnly ? 'text' : 'JSON'}`)
  console.log(`Vosk debug level     : ${debug}`)
  console.log()
  console.log(`load model latency   : ${getTimer('loadModel')}ms`)
  console.log(`transcript latency   : ${getTimer('transcript')}ms`)
  console.log(`transcript text      : ${sentences.join(' ')}`)
  console.log()

}  


/**
 * @function main
 * unit test
 */
async function main() {

  // get command line arguments 
  const { args } = getArgs()
  const { modelDirectory, audioFile, grammar, sampleRate, alternatives, textOnly, tableevents, objectevents, debug } = checkArgs(args)
  const words = ! textOnly

  // set the vosk log level to silence 
  logLevel(debug) 

  setTimer('loadModel')

  // load in memory a Vosk directory model
  const model = loadModel(modelDirectory)

  setTimer('transcript')
  const startSentenceTimer = unixTimeMsecs()

  let transcriptEvents
  const sentences = [] 
  const results = [] 

  // speech recognition from an audio file
  try {
    transcriptEvents = transcriptEventsFromFile(audioFile, model, {grammar, sampleRate, alternatives, words})
  }  
  catch(error) {
    console.error(error) 
  }  

  transcriptEvents.on(PARTIAL_RESULT_EVENT, data => {
    
    if ( !textOnly ) { 

      const dataItem = { 
        time: unixTimeMsecs() - startSentenceTimer,
        event: PARTIAL_RESULT_EVENT,
        data
      }

      results.push(dataItem) 

    }  
  
  })  

  transcriptEvents.on(END_OF_SPEECH_EVENT, data => {
    
    sentences.push(data.text) 
    
    if ( ! textOnly ) {
      
      const dataItem = { 
        time: unixTimeMsecs() - startSentenceTimer,
        event: END_OF_SPEECH_EVENT,
        data
      }

      results.push(dataItem) 

    }  
  
  })  
  
  transcriptEvents.on(FINAL_RESULT_EVENT, data => {
    
    sentences.push(data.text) 

    if ( textOnly ) {
      console.log(sentences.join(' '))
    }  
    else {

      const dataItem = { 
        time: unixTimeMsecs() - startSentenceTimer,
        event: FINAL_RESULT_EVENT,
        data
      }

      results.push(dataItem) 

      if ( !textOnly )
        statistics(modelDirectory, audioFile, grammar, sampleRate, alternatives, textOnly, tableevents, objectevents, debug, sentences)

      if ( !textOnly && tableevents)
        printResultsAsTable(results)

      if ( !textOnly && objectevents)
        console.log(printObject(results, {breakLength:80}))

    }
  
  })  

  // free the runtime model
  freeModel(model)

}


if (require.main === module) 
  main()

module.exports = { 
  logLevel,
  loadModel,
  transcriptFromBuffer,
  transcriptEventsFromBuffer,
  transcriptFromFile,
  transcriptEventsFromFile,
  //transcript: transcriptFromFile, // alias
  freeModel
}

